modules/sfp_massdns.py

```diff
@@ -0,0 +1,406 @@
+# -*- coding: utf-8 -*-
+# -------------------------------------------------------------------------------
+# Name:         sfp_dnsbrute
+# Purpose:      SpiderFoot plug-in for attempting to resolve through brute-forcing
+#               common hostnames.
+#
+# Author:      Steve Micallef <steve@binarypool.com>, TheTechromancer
+#
+# Created:     05/19/2021
+# Copyright:   (c) Steve Micallef 2021
+# Licence:     GPL
+# -------------------------------------------------------------------------------
+
+import re
+import json
+import random
+import tempfile
+import subprocess
+import dns.resolver
+from pathlib import Path
+from spiderfoot import SpiderFootEvent, SpiderFootPlugin
+
+
+class sfp_massdns(SpiderFootPlugin):
+
+    meta = {
+        'name': "MassDNS",
+        'summary': "Attempts to identify hostnames through brute-forcing common names and iterations.",
+        'flags': [""],
+        'useCases': ["Footprint", "Investigate"],
+        'categories': ["DNS"]
+    }
+
+    # Default options
+    opts = {
+        "domainonly": True,
+        "numbermutation": True,
+        "alphamutation": True,
+        "large_wordlist": False,
+        "concurrent_resolvers": 1000
+    }
+
+    # Option descriptions
+    optdescs = {
+        "domainonly": "Only brute-force subdomains for the main target (non-recursive).",
+        "numbermutation": "For any host found, increment/decrement existing numbers (if any) and try appending 1, 01, 001, -1, -01, -001, 2, 02, etc. (up to 10)",
+        "alphamutation": "For any host found, try common mutations such as -test, -old, etc.",
+        "large_wordlist": "Use a 1.2M wordlist instead of the default 110K. Takes roughly 20 minutes at 1000 threads.",
+        "concurrent_resolvers": "Maximum concurrent lookup threads. Bandwidth cost is ~1Mbps per 100 resolvers."
+    }
```

Comment:
Usually modules which leverage third-party tools on the host require the user to specify the file system path in a module option in case the tool is not available in the system environment `PATH`. There's a few reasons to do this, even though this is arguably an environment issue outside of SpiderFoot. @smicallef will likely have an opinion here.

On Windows tools usually aren't installed into the `PATH` by default and often won't add their directory to `PATH`.

There's a security concern due to not using the absolute path. Although only relevant when the system is configured with an insecure `PATH`.

There's a usability concern as there's no error handling in instances where the tool does not exist, making every invocation of the module pointless. Allowing the user to specify a path allows the following error handling:

```python
        if not self.opts['shuffledns_path']:
            self.sf.error("You enabled sfp_tool_massdns but did not set a path to shuffledns!")
            self.errorState = True
            return
```

This will also resolve the failing tests by preventing the test from attempting to execute a file which does note exist:

```
/opt/hostedtoolcache/Python/3.7.11/x64/lib/python3.7/subprocess.py:1551: FileNotFoundError
```

Reply:
The paths are now autodetected via `shutil.which()` if they are left blank in the settings. Checks were added in `handleEvent` to abort if no executable is present.